test/integration/integration.cc

```diff
@@ -699,6 +699,72 @@ void BaseIntegrationTest::testRetry(Http::CodecClient::Type type) {
        [&]() -> void { fake_upstream_connection->waitForDisconnect(); }});
 }
 
+void BaseIntegrationTest::testGrpcRetry() {
+  IntegrationCodecClientPtr codec_client;
+  FakeHttpConnectionPtr fake_upstream_connection;
+  IntegrationStreamDecoderPtr response(new IntegrationStreamDecoder(*dispatcher_));
+  FakeStreamPtr request;
+  Http::TestHeaderMapImpl response_trailers{{"response1", "trailer1"}, {"grpc-status", "0"}};
+  executeActions(
+      {[&]() -> void { codec_client =
+                       makeHttpConnection(lookupPort("http"), Http::CodecClient::Type::HTTP2); },
+       [&]() -> void {
+         Http::TestHeaderMapImpl request_trailers{{"request1", "trailer1"}, {"request2", "trailer2"}};
+         Http::StreamEncoder* request_encoder;
+         request_encoder =
+             &codec_client->startRequest(Http::TestHeaderMapImpl{{":method", "POST"},
+                                                                 {":path", "/test/long/url"},
+                                                                 {":scheme", "http"},
+                                                                 {":authority", "host"},
+                                                                 {"x-forwarded-for", "10.0.0.1"},
+                                                                 {"x-envoy-retry-grpc-on", "cancelled"}},
+                                         *response);
+         codec_client->sendData(*request_encoder, 1024, false);
+         codec_client->sendTrailers(*request_encoder, request_trailers);
+       },
+       [&]() -> void {
+         fake_upstream_connection = fake_upstreams_[0]->waitForHttpConnection(*dispatcher_);
+       },
+       [&]() -> void { request = fake_upstream_connection->waitForNewStream(); },
+       [&]() -> void { request->waitForEndStream(*dispatcher_); },
+       [&]() -> void {
+         request->encodeHeaders(Http::TestHeaderMapImpl{{":status", "200"}, {"grpc-status", "1"}}, false);
+       },
+       [&]() -> void {
+         if (fake_upstreams_[0]->httpType() == FakeHttpConnection::Type::HTTP1) {
```

Comment:
Should we just exclude this from HTTP1 upstream tests? In general you can't proxy gRPC to arbitrary H1 backends I think (certainly not bidi streaming, also you need to get grpc-status into trailers, etc.)